packages/fxa-auth-server/test/local/payments/paypal-processor.js

```diff
@@ -0,0 +1,351 @@
+    Container.set(PayPalHelper, mockPaypalHelper);
+    processor = new PaypalProcessor(mockLog, mockConfig, 1, 1);
+  });
+
+  afterEach(() => {
+    sandbox.reset();
+  });
+
+  describe('constructor', () => {
+    it('sets log, graceDays, retryAttemps, stripe and paypalHelpers', () => {
+      const paypalProcessor = new PaypalProcessor(mockLog, mockConfig, 1, 1);
+      assert.strictEqual(paypalProcessor.log, mockLog);
+      assert.equal(paypalProcessor.graceDays, 1);
+      assert.equal(paypalProcessor.maxRetryAttempts, 1);
+      assert.strictEqual(paypalProcessor.stripeHelper, mockStripeHelper);
+      assert.strictEqual(paypalProcessor.paypalHelper, mockPaypalHelper);
+    });
+  });
+
+  describe('inGracePeriod', () => {
+    it('returns true within grace period', () => {
+      const invoice = deepCopy(unpaidInvoice);
+      const hoursAgo = new Date();
+      hoursAgo.setHours(hoursAgo.getHours() - 20);
+      invoice.created = hoursAgo.getTime() / 1000;
+      const result = processor.inGracePeriod(invoice);
+      assert.isTrue(result);
+    });
+
+    it('returns false when outside of grace period', () => {
+      const invoice = deepCopy(unpaidInvoice);
+      const twoDaysAgo = new Date();
+      twoDaysAgo.setDate(twoDaysAgo.getDate() - 2);
+      invoice.created = twoDaysAgo.getTime() / 1000;
+      const result = processor.inGracePeriod(invoice);
+      assert.isFalse(result);
+    });
+  });
+
+  describe('cancelInvoiceSubscription', () => {
+    it('marks invoice and cancels subscription', async () => {
+      mockStripeHelper.markUncollectible = sandbox.fake.resolves({});
+      mockStripeHelper.cancelSubscription = sandbox.fake.resolves({});
+      const result = await processor.cancelInvoiceSubscription(paidInvoice);
+      assert.deepEqual(result, [{}, {}]);
+      sinon.assert.calledOnceWithExactly(
+        mockStripeHelper.markUncollectible,
+        paidInvoice
+      );
+      sinon.assert.calledOnceWithExactly(
+        mockStripeHelper.cancelSubscription,
+        paidInvoice.subscription
+      );
+    });
+  });
+
+  describe('ensureAccurateAttemptCount', () => {
+    it('does nothing if the attempts match', async () => {
+      mockStripeHelper.getPaymentAttempts = sandbox.fake.returns(1);
+      mockStripeHelper.updatePaymentAttempts = sandbox.fake.resolves({});
+      await processor.ensureAccurateAttemptCount(unpaidInvoice, [{}]);
+      sinon.assert.notCalled(mockStripeHelper.updatePaymentAttempts);
+    });
+
+    it('updates the attempts if they do not match', async () => {
+      const invoice = deepCopy(unpaidInvoice);
+      mockStripeHelper.getPaymentAttempts = sandbox.fake.returns(2);
+      mockStripeHelper.updatePaymentAttempts = sandbox.fake.resolves({});
+      await processor.ensureAccurateAttemptCount(invoice, [{}]);
+      sinon.assert.calledOnceWithExactly(
+        mockStripeHelper.updatePaymentAttempts,
+        invoice,
+        1
+      );
+      sandbox.reset();
```

Comment:
I think a `mockStripeHelper.updatePaymentAttempts.resetHistory()` would also work here.

Reply:
It would, yes. As long as that one is reset it will pass.